engine/ladon/rego/condition/label_selector.rego

```diff
@@ -0,0 +1,389 @@
+        value := labels[label]
+    }
+
+    count(matchLabelPairs - labelPairs) == 0
+}
+
+# labelSelector_expressionMatch returns true if the labels set matches the
+# "matchExpressions" array. Otherwise, returns undefined.
+#
+# Please refer to each function for more details.
+labelSelector_expressionMatch(selector, labels) {
+    labelSelector_expressionMatchIn_AllMatch(selector, labels)
+    labelSelector_expressionMatchNotIn_AllMatch(selector, labels)
+    labelSelector_expressionMatchExists_AllMatch(selector, labels)
+    labelSelector_expressionMatchDoesNotExist_AllMatch(selector, labels)
+}
+
+# labelSelector_expressionMatchIn_AllMatch returns true if _all_ `operator=In`
+# expressions matches with the labels set. Otherwise, returns undefined.
+#
+# An "In" expression matches the labels set if the expression's key exists in it
+# and the value of said key is in the expression's values array.
+labelSelector_expressionMatchIn_AllMatch(selector, labels) {
+    # List of all "In" expressions
+    allInExpressions := [ expr |
+        # Expression is an "In" expression
+        some i
+        expr := selector.matchExpressions[i]
+        expr.operator == "In"
+    ]
+
+    # List of all "In" expressions that matches
+    matchingInExpressions := [ expr |
+        # Expression is an "In" expression
+        some i
+        expr := selector.matchExpressions[i]
+        expr.operator == "In"
+
+        # expr.key must exist in labels
+        _has_key(labels, expr.key)
+
+        # The expr.values array must not be empty
+        count(expr.values) > 0
+
+        # The value of labels[expr.key] must be in the expr.values array
+        _contains(expr.values, labels[expr.key])
+    ]
+
+    count(allInExpressions) == count(matchingInExpressions)
+}
+
+# labelSelector_expressionMatchNotIn_AllMatch returns true if _all_
+# `operator=NotIn` expressions matches with the labels set. Otherwise,
+# returns undefined.
+#
+# A "NotIn" expression matches the labels set if the expression's key exists in
+# it and the value of said key IS NOT in the expression's values array.
+labelSelector_expressionMatchNotIn_AllMatch(selector, labels) {
+    # List of all "NotIn" expressions
+    allNotInExpressions := [ expr |
+        # Expression is an "NotIn" expression
+        some i
+        expr := selector.matchExpressions[i]
+        expr.operator == "NotIn"
+    ]
+
+    # List of all "NotIn" expressions that matches
+    matchingNotInExpressions := [ expr |
+        # Expression is an "NotIn" expression
+        some i
+        expr := selector.matchExpressions[i]
+        expr.operator == "NotIn"
+
+        # Get the metadata.label value, or null if it does not exist
+        labelValue := object.get(labels, expr.key, null)
+
+        # The expr.values array must not be empty
+        count(expr.values) > 0
+
+        # The value of labels[expr.key] must not be in the expr.values array
+        not _contains(expr.values, labels[expr.key])
```

Comment:
Should the `labels[expr.key]` be `labelValue`?

Reply:
Oh you're right. Currently this rego will return false if `labels[expr.key]` is `undefined` (expected behaviour: return true). Will be fixed on the next commit.